compose/decoder/decoder.go

```diff
@@ -0,0 +1,133 @@
+package decoder
+
+import (
+	"io"
+)
+
+type (
+	multiple uint
+
+	FlatReader interface {
+		Read() ([]string, error)
+	}
+
+	StructuredDecoder interface {
+		Decode(interface{}) error
+		More() bool
+	}
+
+	flatReader struct {
+		r      FlatReader
+		header []string
+		more   bool
+	}
+
+	structuredDecoder struct {
+		header []string
+		d      StructuredDecoder
+		buf    []map[string]interface{}
+	}
+)
+
+// flat reader
+func NewFlatReader(r FlatReader) *flatReader {
+	return &flatReader{
+		r:    r,
+		more: true,
+	}
+}
+
+func (dec *flatReader) Get(fnc func([]string)) error {
+	v, err := dec.r.Read()
+	if err == io.EOF {
+		dec.more = false
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	fnc(v)
+	return nil
+}
+
+func (dec *flatReader) Walk(fnc func([]string)) error {
+	for dec.more {
+		if err := dec.Get(fnc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (dec *flatReader) Header() []string {
+	if len(dec.header) > 0 {
+		return dec.header
+	}
+
+	dec.Get(func(rtr []string) {
+		dec.header = rtr
+	})
+
+	return dec.header
+}
+
+// structured decoder
+func NewStructuredDecoder(d StructuredDecoder) *structuredDecoder {
```

Comment:
You'll probably need decoder (for converting one file line to structure, something like `json.Decoder`) AND reader (io.Reader). Then, structured decoder should pass reader through something that knows how to stop at new-line and feed what was read into decoder.